### PR TITLE
feat: customize welcome page

### PR DESCRIPTION
**Pull request by dirien**
Fixes: #550

Hi,

this PR enables potential users to provide their own Welcome page. 

With the flag `--web-template-path=<path>`, you can specify the path to your custom welcome page.

The structure of the folder should be like this:

```bash
web/
  index.html
  xyz.html
  static/
      main.css
      main.js
```

I use `c.HTML` to serve the custom welcome page, this means we have full access to the go-template engine

The rules are: Every file with html ending will be picked up. This is useful if you want to create templates and include them in the main index.html file (see my example). But only the `index.html` will be served!

Static files (css, js or pics) should be saved into the `static` folder. I check the code, if the static folder if present to serve this.

> Currently, their is no support for variables in the template.

If everybody like this, I will extend the enable the helm chart too.

Example:
image

Looking for your feedback!

### CODE REVIEW

File: pkg/chartmuseum/server/multitenant/server.go
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server.go:166]: Unfortunately, by adding the routes here, these pages are no longer protected by auth.

All incoming routes are currently defined in this [file](https://github.com/helm/chartmuseum/blob/main/pkg/chartmuseum/server/multitenant/routes.go) and auth is handled [here](https://github.com/helm/chartmuseum/blob/main/pkg/chartmuseum/router/router.go#L212).

I think we'll have to take a different approach, though I'm not sure the best way forward off the top of my head! 🤔

**Review by cbuto — CHANGES_REQUESTED** [review]
Thanks for the PR! 

I noted one issue around auth that we'll have to figure out.

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:166]: Good point about the auth part! I forgot about this. As we use allow the get for anonymous!

Do you have any suggestions? I am happy to do the rework.

File: web/body.html
Comment (cbuto) [inline web/body.html:1]: 💭 It looks like these are these just example files? If so, I'm not sure if we should commit them here!

File: pkg/chartmuseum/server/multitenant/server.go
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server.go:166]: @dirien I'll have to dig in a bit more but I'm fairly certain we can move the Auth code to be a middleware so it applies to all routes.

Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:166]: Sounds good! I will check this too, 2morrow! thanks for the hint! much appreciate it!

File: web/body.html
Comment (dirien) [reply web/body.html:1]: Yes, should i move this to an folder example or delete it entirely?

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:165]: @cbuto, moved the delivery of potential static files into a handler, behind the auth. 
Tests looked good so far.

File: web/body.html
Comment (cbuto) [reply web/body.html:1]: @dirien I think it would be okay to remove them and use documentation in the README and on chartmuseum.com to explain how to use the feature

File: pkg/chartmuseum/server/multitenant/server.go
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server.go:165]: nice! at first glance, looks good!

File: pkg/chartmuseum/router/match.go
Comment (cbuto) [inline pkg/chartmuseum/router/match.go:128]: Is it possible to update the unit tests to cover this case?

File: pkg/chartmuseum/server/multitenant/handlers.go
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/handlers.go:93]: same here, it be possible to update the unit tests to cover this new handler?

**Review by cbuto — COMMENTED** [review]
thanks for fixing the auth! I think this should be good to go once some basic unit tests are updated/added to cover this new feature. Thanks!!

File: web/body.html
Comment (dirien) [reply web/body.html:1]: ok, @cbuto will remove it. I already have extended the REAMDE. As soon this PR gets merged I will create a PR for the repo holding https://github.com/chartmuseum/www

File: pkg/chartmuseum/router/match.go
Comment (dirien) [reply pkg/chartmuseum/router/match.go:128]: thanks for pointing out, ofc will do!

File: pkg/chartmuseum/server/multitenant/handlers.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/handlers.go:93]: thanks for pointing out, ofc will do!

File: web/body.html
Comment (dirien) [reply web/body.html]: done

File: pkg/chartmuseum/router/match.go
Comment (dirien) [reply pkg/chartmuseum/router/match.go]: done

File: pkg/chartmuseum/server/multitenant/handlers.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/handlers.go:93]: Done

**Comment by dirien** [comment]
Hi @cbuto,

add the tests and removed the folder with the example html files. The Readme has the new feature explained. 

Looking for your feedback :)

File: pkg/chartmuseum/server/multitenant/server.go
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server.go:167]: This will panic if the directory/files don't exist. I'm wondering if the service should continue on even if the files don't exist 🤔

**Review by cbuto — APPROVED** [review]
looks good! thanks for adding the tests!

maybe we can get one more review @nerdeveloper @scbizu?

File: pkg/chartmuseum/server/multitenant/server.go
Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server.go:167]: That might be good since ChartMuseum can function without the static files/web pages.

Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go]: Done, looking for your feedback.

Comment (cbuto) [reply pkg/chartmuseum/server/multitenant/server.go]: Looks good!

**Review by scbizu — CHANGES_REQUESTED** [review]
Generally LGTM , adds some small code suggestions.

**Comment by scbizu** [comment]
@cbuto  I am coming ~ 

@dirien Thanks for your creative PR  ~

File: pkg/chartmuseum/server/multitenant/handlers_test.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/handlers_test.go:14]: thanks, don't know whay my IntellJ did not pick this up! 😅

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go]: good point! Moved to Readdir with first depth!

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go]: yes reworked the fn. Thanks

**Comment by dirien** [comment]
HI @scbizu, thanks for the input. 

I changed the parts! Looking forward for your fedback!

File: pkg/chartmuseum/server/multitenant/server.go
Comment (scbizu) [inline pkg/chartmuseum/server/multitenant/server.go:210]: Put the error check here ? But great choose for using Readdir , should make sense here .

**Review by scbizu — CHANGES_REQUESTED** [review]
@dirien Need some more small changes here XD

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:210]: ups! the error check was under the `range`. Moved it up. Thanks for spotting!

**Comment by dirien** [comment]
@scbizu done. moved the error check above the `range`.

File: pkg/chartmuseum/server/multitenant/server.go
Comment (scbizu) [inline pkg/chartmuseum/server/multitenant/server.go:207]: log error here to report whether if check failed is due to some filesystem issue is better ?

File: pkg/chartmuseum/server/multitenant/server.go
Comment (scbizu) [inline pkg/chartmuseum/server/multitenant/server.go:211]: The same .

**Review by scbizu — CHANGES_REQUESTED** [review]
add some error log is better ?

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:207]: Done, added error log.

File: pkg/chartmuseum/server/multitenant/server.go
Comment (dirien) [reply pkg/chartmuseum/server/multitenant/server.go:212]: np, added error log.

**Review by scbizu — APPROVED** [review]

**Comment by scbizu** [comment]
LGTM , merged .

**Comment by dirien** [comment]
Thanks a lot @scbizu and @cbuto

**Comment by cbuto** [comment]
Thanks for working on this @dirien! 🎉

**Pull request merged**
